core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1079,109 +1103,97 @@ private[spark] class BlockManager(
-            logTrace(s"Trying to replicate $blockId of ${data.size} bytes to $peer")
-            blockTransferService.uploadBlockSync(
-              peer.host,
-              peer.port,
-              peer.executorId,
-              blockId,
-              new NettyManagedBuffer(data.toNetty),
-              tLevel,
-              classTag)
-            logTrace(s"Replicated $blockId of ${data.size} bytes to $peer in %s ms"
-              .format(System.currentTimeMillis - onePeerStartTime))
-            peersReplicatedTo += peer
-            peersForReplication -= peer
-            replicationFailed = false
-            if (peersReplicatedTo.size == numPeersToReplicateTo) {
-              done = true  // specified number of peers have been replicated to
+    val numPeersToReplicateTo = level.replication - 1
+
+    @tailrec def replicateBlock(
+      numFailures: Int,
+      peersForReplication: Seq[BlockManagerId],
+      peersReplicatedTo: Set[BlockManagerId],
+      peersFailedToReplicateTo: Set[BlockManagerId]): Set[BlockManagerId] = {
+
+      if (numFailures > maxReplicationFailures
+                || peersForReplication.isEmpty
+                || peersReplicatedTo.size == numPeersToReplicateTo) {
+        // This selection of a peer and replication is continued in a loop until one of the
+        // following 3 conditions is fulfilled:
+        // (i) specified number of peers have been replicated to
+        // (ii) too many failures in replicating to peers
+        // (iii) no peer left to replicate to
+        peersReplicatedTo
+      } else {
+        val peer = peersForReplication.head
+        val (updatedNumFailures, updatedPeers, updatedReplicatedPeers, updatedFailedPeers) = try {
+          val onePeerStartTime = System.currentTimeMillis
+          logTrace(s"Trying to replicate $blockId of ${data.size} bytes to $peer")
+          blockTransferService.uploadBlockSync(
+            peer.host,
+            peer.port,
+            peer.executorId,
+            blockId,
+            new NettyManagedBuffer(data.toNetty),
+            tLevel,
+            classTag)
+          logTrace(s"Replicated $blockId of ${data.size} bytes to $peer" +
+            s" in ${System.currentTimeMillis - onePeerStartTime} ms")
+          // the block was replicated, lets update state and move ahead
+          (numFailures,
+            peersForReplication.tail,
+            peersReplicatedTo + peer,
+            peersFailedToReplicateTo)
+        } catch {
+          case e: Exception =>
+            logWarning(s"Failed to replicate $blockId to $peer, failure #$numFailures", e)
+            val updatedFailedPeers = peersFailedToReplicateTo + peer
+            // we have a failed replication, so we get the list of peers again
+            // we don't want peers we have already replicated to and the ones that
+            // have failed previously
+            val filteredPeers = getPeers(true).filter{p =>
```

Comment:
Maybe as below,  (there is a good exmaple here, [scala-style-guide#pattern-matching](https://github.com/databricks/scala-style-guide#pattern-matching))
```scala
val filteredPeers = getPeers(true).filter { p =>
```